src/main/scala/is/hail/io/reference/FASTAReader.scala

```diff
@@ -0,0 +1,185 @@
+package is.hail.io.reference
+
+import java.io.{ObjectInputStream, ObjectOutputStream}
+import java.util
+import java.util.Map.Entry
+
+import htsjdk.samtools.reference.{FastaSequenceIndex, FastaSequenceIndexCreator, ReferenceSequenceFile, ReferenceSequenceFileFactory}
+import is.hail.HailContext
+import is.hail.utils._
+import is.hail.variant.{RGBase, ReferenceGenome, Locus}
+import org.apache.hadoop
+
+import scala.language.postfixOps
+import scala.sys.process._
+
+class SerializableReferenceSequenceFile(val hConf: SerializableHadoopConfiguration, val fastaFile: String, val indexFile: String) extends Serializable {
+  @transient var value_ : ReferenceSequenceFile = _
+
+  def value: ReferenceSequenceFile = {
+    if (value_ == null) {
+      val localFastaFile = FASTAReader.getLocalFastaFileName(hConf, fastaFile, indexFile)
+      value_ = ReferenceSequenceFileFactory.getReferenceSequenceFile(new java.io.File(localFastaFile))
+    }
+    value_
+  }
+}
+
+object FASTAReader {
+  @volatile var localFastaFiles = Map.empty[String, String]
+
+  def getLocalIndexFileName(fastaFile: String): String =
+    ReferenceSequenceFileFactory.getFastaIndexFileName(new java.io.File(fastaFile).toPath).toString
+
+  def getLocalFastaFileName(hConf: SerializableHadoopConfiguration, fastaFile: String, indexFile: String): String = {
+    synchronized {
+      if (!localFastaFiles.contains(fastaFile)) {
+        val localFile = FASTAReader.setup(hConf, fastaFile, indexFile)
+        localFastaFiles += fastaFile -> localFile
+      }
+      localFastaFiles(fastaFile)
```

Comment:
Let me start with: This looks mostly correct. If you changed the access modifier on `localFastaFiles` to `private[this]` then I would say it's 100% correct because all accesses to `localFastaFiles` is protected by `FASTAReader`'s intrinsic lock.

---

Now, one issue that people will gripe about with your implementation is that you unnecessarily look up `fastaFile` in the map. In Java, folks would write:

```java
String file = localFastaFiles.get(fastaFile); // returns null if the key isn't present
if (file == null) {
  file = FASTAReader.setup(hConf, fastaFile, indexFile);
  localFastaFiles.put(fastaFile, file);
}
return file;
```

Unfortunately, the most natural, equivalent Scala code would return an `Option`, thus allocating, thus being totally the worst thing ever.

---

There's also a performance issue here. `synchronized` will block everyone from accessing `localFastaFiles`. However, you could imagine that what you really need to block is anyone accessing the particular key that you're running `setup` on. There are special concurrent maps designed to solve exactly this problem.

Modern Java (i.e. Java 8 or 9) would probably advocate the following pattern which has the advantage of not locking the entire map:

```java
private ConcurrentMap<String, String> localFastaFiles = // ...
// ...
localFastaFiles.computeIfAbsent(
    fastaFile,
    _ -> FASTAReader.setup(hConf, fastaFile, indexFile))
```

(I mention the Java way because Java is more or less a programmer lingua franca at this point)

The equivalent in Scala is actually non-allocating this time:

```scala
private[this] val localFastaFiles: concurrent.Map[String, String] = new concurrent.TrieMap()
// ...
def getLocalFastaFileName(...): String =
  localFastaFiles.getOrElseUpdate(fastaFile, FASTAReader.setup(hConf, fastaFile, indexFile))
```

---

I think we should go with the TrieMap approach. It's more performant (since TrieMap is extra special smart with its locking strategy), more obviously correct (no `volatile`s to reason about, no `synchronized`, etc.), and less code to maintain.

NB: It's good practice to use the strictest possible access modifier. There's a variety of issues that just disappear when things aren't accessible to other modules / classes.